ttnn/cpp/ttnn/operations/ccl/all_gather_matmul/device/all_gather_matmul_op.cpp

```diff
@@ -0,0 +1,167 @@
+
+    // Return the AllGatherMatmul program with callbacks
+    return all_gather_matmul_multi_core_with_workers(input_tensors.at(0), output_tensors.at(0), output_tensors.at(2), this->all_gather_struct.dim, this->all_gather_struct.num_links, this->all_gather_struct.ring_size, this->all_gather_struct.ring_index, this->all_gather_struct.receiver_device_id, this->all_gather_struct.sender_device_id, this->all_gather_struct.topology, this->all_gather_core_grid_offset);
+}
+
+namespace operations {
+namespace ccl {
+
+std::vector <ttnn::Tensor> all_gather_matmul(
+    const ttnn::Tensor& input_tensor,
+    const ttnn::Tensor& weight_tensor,
+    const uint32_t dim,
+    const CoreCoord all_gather_core_grid_offset,
+    const uint32_t num_links,
+    const std::optional<MemoryConfig>& memory_config,
+    const bool transpose_a,
+    const bool transpose_b,
+    const std::optional<const DataType> dtype,
+    const std::optional<const tt::operations::primary::MatmulProgramConfig> program_config,
+    const std::optional<const std::string>& activation,
+    const std::optional<const DeviceComputeKernelConfig> compute_kernel_config,
+    const std::optional<const ttnn::CoreGrid> core_grid
+) {
+
+    TT_FATAL(std::getenv("TT_METAL_SLOW_DISPATCH_MODE") == nullptr, "This op is only supported for Fast Dispatch");
+
+    auto devices = input_tensor.get_workers();
+    std::vector<Tensor> output_tensors = {ttnn::Tensor(operation::get_workers_for_op_output({input_tensor, weight_tensor})),
+                                            ttnn::Tensor(operation::get_workers_for_op_output({input_tensor, weight_tensor})),
+                                            ttnn::Tensor(operation::get_workers_for_op_output({input_tensor, weight_tensor}))};
+    std::vector<std::optional<const ttnn::Tensor>> optional_input_tensors = {std::nullopt};
+
+
+    operation::launch_op(
+        [dim, all_gather_core_grid_offset, num_links, memory_config, transpose_a, transpose_b, dtype, program_config, activation, compute_kernel_config, core_grid, devices](
+            const std::vector<Tensor>& input_tensors,
+            const std::vector<std::optional<const ttnn::Tensor>>& optional_input_tensors,
+            const std::vector<std::optional<Tensor>>& optional_output_tensors) mutable -> std::vector<Tensor> {
+
+            const auto& input_tensor = input_tensors.at(0);
+            const auto& weight_tensor = input_tensors.at(1);
+            uint32_t num_devices = devices.size();
+
+            /* All Gather stuff */
+            uint32_t device_index = 0; // Initialize device index
+            uint32_t receiver_device_id = 0; // Initialize receiver device ID
+            uint32_t sender_device_id = 0; // Initialize sender device ID
+
+            for (uint32_t i = 0; i < num_devices; ++i) {
```

Comment:
this is taken from the launch op from all_gather. I've made changes now so that the code isn't duplicated between the launch op in all_gather and the launch op in all_gather_matmul.